### PR TITLE
8305744: (ch) InputStream returned by Channels.newInputStream should have fast path for ReadbleByteChannel/WritableByteChannel

### PR DESCRIPTION
**Pull request by mkarg**
This sub-issue defines the work to be done to implement JDK-8265891 solely for the particular case of ReadableByteChannel-to-WritableByteChannel transfers, where neither Channel is a FileChannel, but including special treatment of SelectableByteChannel.

*Without* this optimization, the JRE applies a loop where a temporary 16k byte array on the Java heap is used to transport the data between both channels. This implies that data is first transported from the source channel into the Java heap and then from the Java heap to the target channel. For most Channels, as these are NIO citizen, this implies two transfers between Java heap and OS resources, which is rather time consuming. As the data is in no way modified by transferTo(), the temporary allocation of valuable heap memory is just as unnecessary as both transfers to and from the heap. It makes sense to prevent the use of Java heap memory.

This optimization omits the byte array on the Java heap, effectively omitting the transfers to and from that array in turn. Instead, the transfer happens directly from Channel to Channel reusing a ByteBuffer taken from the JRE's internal buffer management utility which also has a size of 16k. Both measures make the transfer more efficient, as just the very essential resources are used, but not more.

Using JMH on an in-memory data transfer using custom Channels (to prevent interception by other optimizations already existing in transferTo()) it was possible to measure a 2,5% performance gain. While this does not sound much, just as the spared 16k of Java heap doesn't, we need to consider that this also means sparing GC pressure, power consumption and CO2 footprint, and we need to consider that it could be higher when using different (possibly third-party) kinds of custom Channels (as "real I/O" between heap and OS-resources is way slower than the RAM-to-RAM benchmark applied by me). As the change does not induce new, extraordinary risks compared to the existing source code, I think it is worth gaining that 2,5+ percent. Possibly there might exist some kind of scalable Java-implemented cloud service that is bound to exactly this loop and that transports very heavy traffic, those particular 2,5+ percent could be a huge amount of kWh or CO2 in absolute numbers. Even if not now, there might be in future.

---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8305744](https://bugs.openjdk.org/browse/JDK-8305744): (ch) InputStream returned by Channels.newInputStream should have fast path for ReadbleByteChannel/WritableByteChannel (**Sub-task** - P4)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/13387/head:pull/13387` \
`$ git checkout pull/13387`

Update a local copy of the PR: \
`$ git checkout pull/13387` \
`$ git pull https://git.openjdk.org/jdk.git pull/13387/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 13387`

View PR using the GUI difftool: \
`$ git pr show -t 13387`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/13387.diff">https://git.openjdk.org/jdk/pull/13387.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/13387#issuecomment-1500218245)

### CODE REVIEW

File: src/java.base/share/classes/sun/nio/ch/ChannelInputStream.java [hunk 1]
```diff
@@ -307,6 +339,28 @@ private static long transfer(ReadableByteChannel src, FileChannel dst) throws IO
...
+            for (int bytesRead = src.read(bb); bytesRead > -1; bytesRead = src.read(bb)) {
+                bb.flip();
+                while (bb.hasRemaining())
+                    dst.write(bb);
+                bb.clear();
+                bytesWritten += bytesRead;
+            }
+            return bytesWritten;
+        } finally {
+            Util.releaseTemporaryDirectBuffer(bb);
```
Comment (ecki) [inline src/java.base/share/classes/sun/nio/ch/ChannelInputStream.java:360]: Maybe put that in a finally.

Comment (ecki) [reply src/java.base/share/classes/sun/nio/ch/ChannelInputStream.java:360]: (Ah sorry it is, missed it)

File: src/java.base/share/classes/sun/nio/ch/ChannelInputStream.java [hunk 2]
```diff
@@ -265,11 +265,43 @@ public long transferTo(OutputStream out) throws IOException {
...
+
+            if (rbc instanceof SelectableChannel rsc) {
+                synchronized (rsc.blockingLock()) {
+                    if (!rsc.isBlocking())
+                        throw new IllegalBlockingModeException();
+                    if (wbc instanceof SelectableChannel wsc) {
+                        synchronized (wsc.blockingLock()) {
+                            if (!wsc.isBlocking())
+                                throw new IllegalBlockingModeException();
+                            return transfer(rbc, wbc);
```
Comment (AlanBateman) [inline src/java.base/share/classes/sun/nio/ch/ChannelInputStream.java:281]: This introduces the potential for deadlock as tock ordering is not specified for the transferXXX methods. Note that this same issue is problematic for some of the optimisations at the lower level in FileChannel too. We'll need to think about this topic.

File: src/java.base/share/classes/sun/nio/ch/ChannelInputStream.java [hunk 3]
```diff
@@ -265,11 +265,43 @@ public long transferTo(OutputStream out) throws IOException {
             return transfer(rbc, fc);
         }
 
+        // ReableByteChannel -> WritableByteChannel
```
Comment (AlanBateman) [inline src/java.base/share/classes/sun/nio/ch/ChannelInputStream.java:268]: Typo, should be "ReadbleByteChannel"

File: src/java.base/share/classes/sun/nio/ch/ChannelInputStream.java [hunk 4]
```diff
@@ -265,11 +265,43 @@ public long transferTo(OutputStream out) throws IOException {
...
+        if (out instanceof ChannelOutputStream cos) {
```
Comment (AlanBateman) [inline src/java.base/share/classes/sun/nio/ch/ChannelInputStream.java:269]: I think it might be better to change the code at L254 to:

```
        if (out instanceof ChannelOutputStream cos) {
            ReadableByteChannel rbc = ch;
            WritableByteChannel wbc = cos.channel();

            // FileChannel -> WritableByteChannel

            // ...
        }
```

That puts the special casing for an output stream backed by a ChannelOutputStream in one place.

**Review by AlanBateman — COMMENTED** [review]
It would be helpful if there were some benchmarks or something to that shows which cases this helps. The current description isn't very clear on which case there is 2.5% benefit.

File: src/java.base/share/classes/sun/nio/ch/ChannelInputStream.java (hunk 3 above)
Comment (mkarg) [reply src/java.base/share/classes/sun/nio/ch/ChannelInputStream.java]: > Typo, should be "ReadbleByteChannel"

[sic] / pun intended 😉? Thank you, good catch! I fixed it in 0a7e020928f010c42477df5f04c69b915182bbac.

File: src/java.base/share/classes/sun/nio/ch/ChannelInputStream.java (hunk 4 above)
Comment (mkarg) [reply src/java.base/share/classes/sun/nio/ch/ChannelInputStream.java]: Agreed and done in 7039304d0ab75fdf7d400a3e8d0f60cd9d535ad6. Funny anecdote: Originally I had it this way but then modified it to simplify the review.

**Comment by mkarg** [comment]
> It would be helpful if there were some benchmarks or something to that shows which cases this helps. The current description isn't very clear on which case there is 2.5% benefit.

I assume the *biggest* benefit would occur when running on special I/O hardware with custom OS drivers and custom Channel implementations (like forwarding analog samples from an ADC input to a DAC output, e. g. in audio devices). As I do not own such a device, and as the *biggest* benefit should not be the basis for decision-making, I haven choosen the opposite for my benchmark: I measured the code that might benefit *the least* from the optimization: Java-Heap-to-Java-Heap. Still the result is ~2.5% benefit.
```java
@Benchmark
public void ChannelInputStream_transferTo() throws Throwable {
    var is = Channels.newInputStream(Channels.newChannel(new ByteArrayInputStream(new byte[1024 * 1024 * 1024])));
    var os = Channels.newOutputStream(Channels.newChannel(new ByteArrayOutputStream(1024 * 1024 * 1024)));
    is.transferTo(os);
}
```
```
Sandbox.ChannelInputStream_transferTo  thrpt   25  3,794 ┬▒ 0,095  ops/s -- incl. optimization (i. e. using ByteBuffer)
Sandbox.ChannelInputStream_transferTo  thrpt   25  3,703 ┬▒ 0,098  ops/s -- excl. optimization (i. e. using byte[])
```

**Comment by AlanBateman** [comment]
> I assume the _biggest_ benefit would occur when running on special I/O hardware with custom OS drivers and custom Channel implementations (like forwarding analog samples from an ADC input to a DAC output, e. g. in audio devices). As I do not own such a device, and as the _biggest_ benefit should not be the basis for decision-making, I haven choosen the opposite for my benchmark: I measured the code that might benefit _the least_ from the optimization: Java-Heap-to-Java-Heap. Still the result is ~2.5% benefit.

I think the proposed changes are for second order scenarios. It's for cases like an input stream on a Pipe.SourceChannel and transferTo called to transfer all bytes to an output stream on a writable byte channel, or the cases like an input stream on some arbitrary readable byte channel and transferTo is called to transfer all bytes to an output stream on a Pipe.SinkChannel. It has no impact on common cases where the input stream or target is backed by a FileChannel or SocketChannel.  I'm not opposed to trying to improve more of these cases, it's just that it's adding more code and complexity for scenarios that may not be too common.

**Comment by mkarg** [comment]
>It has no impact on common cases where the input stream or target is backed by a FileChannel or SocketChannel. I'm not opposed to trying to improve more of these cases, it's just that it's adding more code and complexity for scenarios that may not be too common.

This is correct. My current PR is intended to improve exactly those scenarios where *no* file or socket is involved, as file and socket were optimized already in my earlier PRs over the last years. Nevertheless, I do not see that the complexity or code size added would be *inappropriate*: All we do here simply is adding a loop-over-ByteBuffer, which is simple and straightforward. The complexity solely stems  from the combinatory numbers of selectable or not selectable channels.

File: src/java.base/share/classes/sun/nio/ch/ChannelInputStream.java (hunk 2 above)
Comment (mkarg) [reply src/java.base/share/classes/sun/nio/ch/ChannelInputStream.java:280]: IIUC the JRE already has the same problem in *other* locations, so wouldn't it make sense to merge the current PR as-is (to gain its benefit already in the upcoming LTS release) and postpone the discussion about possible deadlocks into a separate PR which then covers *all* those existing problematic code locations as a whole? Because I do not see why to withhold *just this* PR due to a problem existing *already* in the JRE.

Comment (AlanBateman) [reply src/java.base/share/classes/sun/nio/ch/ChannelInputStream.java:280]: > IIUC the JRE already has the same problem in _other_ locations, so wouldn't it make sense to merge the current PR as-is (to gain its benefit already in the upcoming LTS release) and postpone the discussion about possible deadlocks into a separate PR which then covers _all_ those existing problematic code locations as a whole? Because I do not see why to withhold _just this_ PR due to a problem existing _already_ in the JRE.

Indeed, the issues of async close of the source/target during transfer, and the potential for deadlock with transfer between bidirectional streams is mostly a lower level concern for methods like FileChannel.transferXXX. It may be that we have to specify an ordering that higher level participants (ChannelInputStream.transferTo) would need to follow.

Comment (mkarg) [reply src/java.base/share/classes/sun/nio/ch/ChannelInputStream.java:280]: I do agree.

**Comment by bplb** [comment]
> Sandbox.ChannelInputStream_transferTo  thrpt   25  3,794 ┬▒ 0,095  ops/s -- incl. optimization (i. e. using ByteBuffer)
> Sandbox.ChannelInputStream_transferTo  thrpt   25  3,703 ┬▒ 0,098  ops/s -- excl. optimization (i. e. using byte[])

If these results are obtained using the default settings, then if I am not mistaken, the following represent the 99.9% confidence intervals:
```
[3.699, 3.889] incl. optimization
[3.605, 3.801] excl. optimization
```
As these intervals overlap, it is not really valid in a statistical sense to definitively state that there is an improvement in throughput. In general to have a reasonable level of certainty that there is in fact improvement, the intervals should not overlap at all and the widths of the intervals should be small compared to the magnitudes of the values being compared. The widths of the intervals are small enough here, but given that the intervals overlap by more than one third one must question the significance of the change.

**Comment by mkarg** [comment]
>...given that the intervals overlap by more than one third one must question the significance of the change.

I see your statistical point (BTW overlap of just one third *is* good enough for me, as it is definitively *not the same* performance), but we are not just statisticians we are also engineers, and share the knowledge that *not executing* code *will* always be faster than *executing* code (I think the only difference we have is that you expect *more* improvement). So given the fact that still there is *some* improvement (the statistics do not imply that there is *no* improvement, it only implies that it is *small*), I really, really would like to contribute this change, still. My points are:
* *Every single* spared CPU cycle is a good thing these days, even if the performance benefit is small for a *single* application, in the end this sums up to lots of spared CPU cycles when looking at the mere sum *all* applications on earth. In a mathematical sense: 1 * n still is n but not 0.
* As a result, the world spares not only CPU cycles, but also spares power consumption, and with this, produces less need for cooling, which in turn spares power consumption. So in the end, in sum we globally spare some tons of CO2 per years. Even if it is only very few tons, it is a benefit for the world climate, gained for rather free. So why *not* doing it?
* I see that we need to trade-off efforts, risk and benefit. For me, the effort from here to target is nearly zero, as the code is already done, is reviewed, is proven functional, and is test-covered. The additional risk of this PR is not much higher than the risk we accepted already with any other optimization we already did in the area of `transferTo` in the past years. For me, each single spared CPU cycle / each single spared ton of CO2 counts much, much more than anything else.

So I plea to accept this PR, even if not statistically evident! :-)

**Comment by mkarg** [comment]
> Beyond the merits of the PR itself, beware the [sunk cost falacy](https://developerexperience.io/articles/sunk-cost). And consider the future maintenance cost of the additional code, its complexity, and interactions.

I did, and still believe that it is worth it.

**Comment by bplb** [comment]
> ```java
> @Benchmark
> public void ChannelInputStream_transferTo() throws Throwable {
>     var is = Channels.newInputStream(Channels.newChannel(new ByteArrayInputStream(new byte[1024 * 1024 * 1024])));
>     var os = Channels.newOutputStream(Channels.newChannel(new ByteArrayOutputStream(1024 * 1024 * 1024)));
>     is.transferTo(os);
> }
> ```

The above benchmark method includes the construction of 6 objects in the measurement. I think it would be more representative of the effect of the change to do something like this:

```
    private static final int SIZE = 1024*1024*1024; 
 
    ByteArrayInputStream in = new ByteArrayInputStream(new byte[SIZE]);
    ByteArrayOutputStream out = new ByteArrayOutputStream(SIZE);
 
    ReadableByteChannel rbc = Channels.newChannel(in);
    WritableByteChannel wbc = Channels.newChannel(out);

    InputStream is = Channels.newInputStream(rbc);
    OutputStream os = Channels.newOutputStream(wbc);

    @Setup
    public void setup() {
        in.mark(SIZE + 1);
    }

    @Benchmark
    public void ChannelInputStream_transferTo() throws Throwable {
        is.transferTo(os);
        in.reset();
        out.reset();
    }
```

**Comment by mkarg** [comment]
> The above benchmark method includes the construction of 6 objects in the measurement. I think it would be more representative of the effect of the change to do something like this:

Correct, the benchmark could be tweaked to proof better results, but as I said in my original posting of that code: "I measured the code that might benefit *the least* from the optimization", as my claim was: It does *never* do harm, so *why not* doing it. :-)

**Comment by bplb** [comment]
> Correct, the benchmark could be tweaked to proof better results

It would be better if the benchmark were so updated.

**Comment by mkarg** [comment]
> It would be better if the benchmark were so updated.

I will change the benchmark as you proposed and run it again.

**Comment by bplb** [comment]
> I will change the benchmark as you proposed and run it again.

Thanks!

**Comment by bplb** [comment]
> @Setup
>     public void setup() {
>         in.mark(SIZE + 1);
>     }

In my benchmark example the above code is unneeded.

**Comment by mkarg** [comment]
# TL;DR

This MR outperforms the master branch's code by one order of magnitude in the best case, but is never slower in the worst case.


# Proof

To proof this claim, Brian's benchmark is used with two different edge cases.

## Lower Bound: ByteArray-Streams

Using Brian's code (see above) I could proof that this MR is always at-least-as fast as the original code.

```
ByteArray-Streams using code from this MR:

Result "eu.sandbox.Sandbox.ChannelInputStream_transferTo":
  2.727 ±(99.9%) 0.096 ops/s [Average]
  (min, avg, max) = (2.526, 2.727, 2.900), stdev = 0.128
  CI (99.9%): [2.631, 2.823] (assumes normal distribution)


ByteArray-Streams using original code from master branch:

Result "eu.sandbox.Sandbox.ChannelInputStream_transferTo":
  2.725 ±(99.9%) 0.103 ops/s [Average]
  (min, avg, max) = (2.551, 2.725, 2.935), stdev = 0.138
  CI (99.9%): [2.622, 2.828] (assumes normal distribution)

```
The reason that no statistical evidence for *improved* performance could be found *in this edge case* is clearly bound to the fact that what we measure here mostly is the time needed by the ByteArray*Streams to transfer bytes between the byte arrays. This needs *very* long time - *much* more time than the actual code change that we actually wanted to benchmark. Copying Gigabytes between the byte arrays is *not* the target of this MR's optimization, but happens in exact the same way in the original code just as in the MR's code, and is (more or less) *solely* what Brian's bechnmark is actually measuring. Hence it is clear, that the statistical result reflects just the fact that copying Gigabytes costs a lot of time, but says rather nothing about the *actual* optimization this MR is about.


## Upper Bound: No-Op-Channels

In fact, what this MR optimizes is not Channel-wrapped ByteArray*Streams (even if Brian's benchmark has proven that they will not work any slower than before), but is -as I explained earlier- custom channels. As `Channel` is a public interface since very long time (starting with Java 1.4), it is possible and likely that custom implementations *do* exist in field use. There could be, for example, channels that use native memory *solely*, even there could be channels that do not transfer the data at all but simple forward buffers from reable channel to writable channel in a hardware-supported way. We do not know what those optimized custom channels actually work like, but as this optimization is about the proof that channel-to-channel transfer is (way) faster than channel-to-bytearray-to-channel transfer (as the original code works like), we actually do not care about the actual implementation of the custom channels. Moreover, we do not care about that, because both benchmark runs (the master code just as the MR's code) will read and write the channels in full length using the same public interface methods, so it is irrelevant what these methods actually do (as long as they do it correctly). Basing on this theory, I have modified Brian's code to not further use byte arrays in those channels, but to use No-Op-Channels. This means, channels that behave *correctly* (i. e. their interface methods work consistently with the spec, and the input channel provides the full length of bytes) but whose implementations do not *copy* any bytes. Using this variation of the benchmark, we now have the means to measure what we actually wanted to measure in the first place: How much faster is the MR's code compared to master's code? Hence: We do not measure anymore how much time the particular *implementation* of an arbitraritly chosen Input/OutputStream needs - again, as this is the same time for all tests as this code is *unchanged*.

Using No-Op-Channels I could proof that the actual modification is highly beneficial, as the MR's code clearly outperforms master's code by one full order of magnitude:
```
No-Op-Channel-Streams using code from this MR:

Result "eu.sandbox.Sandbox.ChannelInputStream_transferTo":
  3403.370 ±(99.9%) 287.694 ops/s [Average]
  (min, avg, max) = (1977.793, 3403.370, 3759.044), stdev = 384.063
  CI (99.9%): [3115.676, 3691.063] (assumes normal distribution)


No-Op-Channel-Streams using code from master:

Result "eu.sandbox.Sandbox.ChannelInputStream_transferTo":
  349.644 ±(99.9%) 27.853 ops/s [Average]
  (min, avg, max) = (240.305, 349.644, 366.043), stdev = 37.182
  CI (99.9%): [321.792, 377.497] (assumes normal distribution)
```
I assume that the statistical evidence is obvious, just as the reasoning is: Not having to move one Gigabyte of data from a ByteBuffer into a byte array and back once per test iteration certainly is (much) faster than doing it. While CPUs and memory paths are super fast these days, the transfer costs still are *not zero*, and one Gigabyte transferred twice per test makes this rather evident. That free CPU, MMU and transfer path resources (and the time needed for the spared GCs) can be used better for parallel tasks (or spared completely to spare CO2).

Below you can find the implementation of the Null-Op-Channels:
```java
	private static NullReadableByteChannel newNullReadableByteChannel(final int size) {
		return new NullReadableByteChannel(size);
	}

	private static class NullReadableByteChannel implements ReadableByteChannel {
		private NullReadableByteChannel(final int size) {
			this.size = size;
			s = size;
		}
		int s, size;

		private boolean isClosed;

		@Override
		public boolean isOpen() {
			return !isClosed;
		}

		@Override
		public void close() throws IOException {
			this.isClosed = true;
		}

		@Override
		public int read(ByteBuffer dst) throws IOException {
			if (isClosed)
				throw new ClosedChannelException();
			if (s <= 0)
				return -1;
			final int p = dst.position();
			final int r = dst.remaining();
			final int c = Math.min(r, s);
			dst.position(p + c);
			s -= c;
			return c;
		}

		public void reset() {
			s = size;
		}
	}

	private static WritableByteChannel newNullWritableByteChannel() {
		return new WritableByteChannel() {
			private boolean isClosed;

			@Override
			public boolean isOpen() {
				return !isClosed;
			}

			@Override
			public void close() throws IOException {
				this.isClosed = true;
			}

			@Override
			public int write(ByteBuffer src) throws IOException {
				if (isClosed)
					throw new ClosedChannelException();
				final int p = src.position();
				final int r = src.remaining();
				final int c = r;
				src.position(p + c);
				return c;
			}			
		};
	}
```

**Comment by mkarg** [comment]
@bplb I would be glad if we could continue with the review of this PR.
